File: lib/python/src/cabincrew_protocol/protocol.py
from dataclasses import dataclass
from typing import Optional, List, Dict, Any, Union
from enum import Enum


@dataclass
class RecordStringAny:
    """Arbitrary metadata. Optional."""

    pass


@dataclass
class PreflightEvidence:
    hash: str = None
    name: str = None
    path: str = None


@dataclass
class ApprovalRequest:
    approval_id: str = None
    reason: str = None
    required_role: str = None
    step_id: str = None
    workflow_id: str = None
    engine_output: Optional[RecordStringAny] = None
    evidence: Optional[List[PreflightEvidence]] = None
    plan_token_hash: Optional[str] = None


@dataclass
class ApprovalResponse:
    approval_id: str = None
    approved: bool = None
    approver: Optional[str] = None
    reason: Optional[str] = None
    timestamp: Optional[str] = None


@dataclass
class Artifact:
    """Canonical artifact interface.
    Defined in schemas/draft/artifact.schema.json
    """
    action: str = None
    """Operation to perform with this artifact (create, update, delete, apply, execute, etc).
    Free-form and engine-defined.
    """
    artifact_type: str = None
    """Type of artifact (file, diff, patch, action, message, etc). Free-form and engine-defined."""

    mime: str = None
    """MIME type describing content."""

    body: Optional[Union[Dict[str, Any], List[Any], str]] = None
    """Inline content for small artifacts.
    Can be string, object, array, or null.
    """
    body_file: Optional[str] = None
    """Indicates an external data file within the artifact directory."""

    metadata: Optional[RecordStringAny] = None
    """Arbitrary metadata. Optional."""

    target: Optional[str] = None
    """Path, resource, or identifier this artifact applies to. Optional."""


@dataclass
class AuditApproval:
    approval_id: Optional[str] = None
    approved: Optional[bool] = None
    approver: Optional[str] = None
    reason: Optional[str] = None
    required_role: Optional[str] = None


@dataclass
class AuditArtifact:
    hash: Optional[str] = None
    name: Optional[str] = None
    path: Optional[str] = None
    role: Optional[str] = None
    size: Optional[float] = None


@dataclass
class AuditEngine:
    engine_id: Optional[str] = None
    error: Optional[str] = None
    receipt_id: Optional[str] = None
    status: Optional[str] = None


@dataclass
class AuditGateway:
    gateway_type: Optional[str] = None
    model: Optional[str] = None
    policy_decision: Optional[str] = None
    request_id: Optional[str] = None
    tool: Optional[str] = None


@dataclass
class AuditIntegrity:
    actual_plan_token: Optional[str] = None
    artifacts_match: Optional[bool] = None
    differences: Optional[List[str]] = None
    expected_plan_token: Optional[str] = None
    plan_token_match: Optional[bool] = None


@dataclass
class PlanArtifactHash:
    hash: str = None
    name: str = None
    size: Optional[float] = None


@dataclass
class PlanToken:
    """Cryptographic binding between a flight-plan and its subsequent take-off.
    Defined in schemas/draft/plan-token.schema.json
    """
    artifacts: List[PlanArtifactHash] = None
    """Per-artifact hashes that contributed to this plan token."""

    created_at: str = None
    """Timestamp when the plan was created (RFC3339)."""

    engine_id: str = None
    """Engine identity that produced this plan."""

    model: str = None
    """AI Model identifier used to generate this plan (e.g. 'gpt-4', 'claude-3').
    Required for provenance.
    """
    protocol_version: str = None
    """Engine protocol version used when this plan was produced."""

    token: str = None
    """Primary plan token identifier, e.g. SHA256 over all plan artifacts + context."""

    workspace_hash: str = None
    """Hash of the workspace state when the plan was created."""


@dataclass
class AuditPolicy:
    decision: Optional[str] = None
    engine: Optional[str] = None
    violations: Optional[List[str]] = None
    warnings: Optional[List[str]] = None


class Severity(Enum):
    CRITICAL = "critical"
    DEBUG = "debug"
    ERROR = "error"
    INFO = "info"
    WARNING = "warning"


@dataclass
class AuditWorkflow:
    mode: Optional[str] = None
    step_id: Optional[str] = None
    workflow_id: Optional[str] = None


@dataclass
class AuditEvent:
    """Canonical schema for all audit log events.
    Defined in schemas/draft/audit-event.schema.json
    """
    event_id: str = None
    """Unique identifier for this audit event."""

    event_type: str = None
    """Free-form event category."""

    timestamp: str = None
    """RFC3339 timestamp of when the event occurred."""

    approval: Optional[AuditApproval] = None
    artifacts: Optional[List[AuditArtifact]] = None
    chain_hash: Optional[str] = None
    """Hash of the previous event in the chain. Allows for ledger-style verification."""

    engine: Optional[AuditEngine] = None
    gateway: Optional[AuditGateway] = None
    integrity_check: Optional[AuditIntegrity] = None
    message: Optional[str] = None
    plan_token: Optional[PlanToken] = None
    """Cryptographic binding between a flight-plan and its subsequent take-off.
    Defined in schemas/draft/plan-token.schema.json
    """
    policy: Optional[AuditPolicy] = None
    severity: Optional[Severity] = None
    signature: Optional[str] = None
    """Cryptographic signature of this event hash."""

    signature_key_ref: Optional[str] = None
    """Reference to the key used for signing (e.g. 'engine-key-1', 'orchestrator-key-prod')."""

    workflow: Optional[AuditWorkflow] = None


@dataclass
class RecordStringAnyClass:
    pass


class Decision(Enum):
    ALLOW = "allow"
    DENY = "deny"
    REQUIRE_APPROVAL = "require_approval"
    WARN = "warn"


@dataclass
class EngineArtifact:
    hash: str = None
    name: str = None
    path: str = None
    role: str = None
    size: Optional[float] = None


@dataclass
class EngineMeta:
    step_id: str = None
    workflow_id: str = None


class Mode(Enum):
    """Execution mode: 'flight-plan' or 'take-off'."""

    FLIGHT_PLAN = "flight-plan"
    TAKE_OFF = "take-off"


@dataclass
class EngineOrchestrator:
    artifacts_salt: Optional[str] = None
    run_index: Optional[float] = None
    workspace_hash: Optional[str] = None


@dataclass
class EngineInput:
    """Input delivered via STDIN or CABINCREW_INPUT_FILE.
    Defined in schemas/draft/engine.schema.json
    """
    meta: EngineMeta = None
    mode: Mode = None
    """Execution mode: 'flight-plan' or 'take-off'."""

    protocol_version: str = None
    allowed_secrets: Optional[List[str]] = None
    config: Optional[Dict[str, Any]] = None
    context: Optional[Dict[str, Any]] = None
    expected_plan_token: Optional[str] = None
    identity_token: Optional[str] = None
    """Ephemeral identity token (e.g. OIDC, JWT) for the workload.
    Preferred over static secrets.
    """
    orchestrator: Optional[EngineOrchestrator] = None
    secrets: Optional[Dict[str, Any]] = None


@dataclass
class EngineMetric:
    name: str = None
    value: float = None
    tags: Optional[Dict[str, Any]] = None


@dataclass
class EngineOutput:
    """Output delivered via STDOUT or CABINCREW_OUTPUT_FILE.
    Defined in schemas/draft/engine.schema.json
    """
    engine_id: str = None
    mode: Mode = None
    protocol_version: str = None
    receipt_id: str = None
    status: str = None
    """Execution status: 'success' or 'failure'."""

    artifacts: Optional[List[EngineArtifact]] = None
    diagnostics: Any = None
    error: Optional[str] = None
    metrics: Optional[List[EngineMetric]] = None
    plan_token: Optional[str] = None
    """SHA256 hash referencing a plan-token.json file."""

    warnings: Optional[List[str]] = None


@dataclass
class GatewayApproval:
    approval_id: Optional[str] = None
    reason: Optional[str] = None
    required_role: Optional[str] = None


@dataclass
class LLMGatewayRule:
    action: str = None
    match: RecordStringAny = None
    metadata: Optional[RecordStringAny] = None


@dataclass
class LLMGatewayPolicyConfig:
    model_routing: Optional[RecordStringAny] = None
    onnx_models: Optional[List[str]] = None
    opa_policies: Optional[List[str]] = None
    rules: Optional[List[LLMGatewayRule]] = None


@dataclass
class LLMGatewayRequest:
    input: RecordStringAny = None
    model: str = None
    request_id: str = None
    timestamp: str = None
    context: Optional[RecordStringAny] = None
    provider: Optional[str] = None
    source: Optional[str] = None


@dataclass
class LLMGatewayResponse:
    decision: Decision = None
    request_id: str = None
    timestamp: str = None
    approval: Optional[GatewayApproval] = None
    gateway_payload: Optional[RecordStringAny] = None
    rewritten_input: Optional[RecordStringAny] = None
    routed_model: Optional[str] = None
    violations: Optional[List[str]] = None
    warnings: Optional[List[str]] = None


@dataclass
class MCPGatewayRule:
    action: str = None
    match: RecordStringAny = None
    metadata: Optional[RecordStringAny] = None


@dataclass
class MCPGatewayPolicyConfig:
    onnx_models: Optional[List[str]] = None
    opa_policies: Optional[List[str]] = None
    rules: Optional[List[MCPGatewayRule]] = None


@dataclass
class MCPGatewayRequest:
    method: str = None
    request_id: str = None
    server_id: str = None
    timestamp: str = None
    context: Optional[RecordStringAny] = None
    params: Optional[RecordStringAny] = None
    source: Optional[str] = None


@dataclass
class MCPGatewayResponse:
    decision: Decision = None
    request_id: str = None
    timestamp: str = None
    approval: Optional[GatewayApproval] = None
    rewritten_request: Optional[RecordStringAny] = None
    violations: Optional[List[str]] = None
    warnings: Optional[List[str]] = None


@dataclass
class PreflightInput:
    engine_output: RecordStringAny = None
    mode: Mode = None
    step_id: str = None
    workflow_id: str = None
    context: Optional[RecordStringAny] = None
    evidence: Optional[List[PreflightEvidence]] = None
    plan_token: Optional[PlanToken] = None
    """Cryptographic binding between a flight-plan and its subsequent take-off.
    Defined in schemas/draft/plan-token.schema.json
    """


@dataclass
class PreflightRequires:
    reason: Optional[str] = None
    role: Optional[str] = None


@dataclass
class PreflightOutput:
    decision: Decision = None
    requires: Optional[PreflightRequires] = None
    violations: Optional[List[str]] = None
    warnings: Optional[List[str]] = None


class State(Enum):
    APPROVED = "APPROVED"
    ARTIFACTS_VALIDATED = "ARTIFACTS_VALIDATED"
    AWAITING_APPROVAL = "AWAITING_APPROVAL"
    COMPLETED = "COMPLETED"
    EXECUTION_COMPLETE = "EXECUTION_COMPLETE"
    FAILED = "FAILED"
    INIT = "INIT"
    PLAN_GENERATED = "PLAN_GENERATED"
    PLAN_RUNNING = "PLAN_RUNNING"
    PREFLIGHT_COMPLETE = "PREFLIGHT_COMPLETE"
    PRE_FLIGHT_RUNNING = "PRE_FLIGHT_RUNNING"
    READY_FOR_TAKEOFF = "READY_FOR_TAKEOFF"
    TAKEOFF_RUNNING = "TAKEOFF_RUNNING"
    TOKEN_CREATED = "TOKEN_CREATED"


@dataclass
class WorkflowState:
    state: State = None
    last_decision: Optional[str] = None
    plan_token_hash: Optional[str] = None
    step_id: Optional[str] = None
    workflow_id: Optional[str] = None


@dataclass
class CabinCrewProtocol:
    any_map: Optional[Dict[str, Any]] = None
    approval_request: Optional[ApprovalRequest] = None
    approval_response: Optional[ApprovalResponse] = None
    artifact: Optional[Artifact] = None
    audit_approval: Optional[AuditApproval] = None
    audit_artifact: Optional[AuditArtifact] = None
    audit_engine: Optional[AuditEngine] = None
    audit_event: Optional[AuditEvent] = None
    audit_gateway: Optional[AuditGateway] = None
    audit_integrity: Optional[AuditIntegrity] = None
    audit_policy: Optional[AuditPolicy] = None
    audit_workflow: Optional[AuditWorkflow] = None
    decision: Optional[Decision] = None
    engine_artifact: Optional[EngineArtifact] = None
    engine_input: Optional[EngineInput] = None
    engine_meta: Optional[EngineMeta] = None
    engine_metric: Optional[EngineMetric] = None
    engine_orchestrator: Optional[EngineOrchestrator] = None
    engine_output: Optional[EngineOutput] = None
    gateway_approval: Optional[GatewayApproval] = None
    llm_gateway_policy_config: Optional[LLMGatewayPolicyConfig] = None
    llm_gateway_request: Optional[LLMGatewayRequest] = None
    llm_gateway_response: Optional[LLMGatewayResponse] = None
    llm_gateway_rule: Optional[LLMGatewayRule] = None
    mcp_gateway_policy_config: Optional[MCPGatewayPolicyConfig] = None
    mcp_gateway_request: Optional[MCPGatewayRequest] = None
    mcp_gateway_response: Optional[MCPGatewayResponse] = None
    mcp_gateway_rule: Optional[MCPGatewayRule] = None
    mode: Optional[Mode] = None
    plan_artifact_hash: Optional[PlanArtifactHash] = None
    plan_token: Optional[PlanToken] = None
    preflight_evidence: Optional[PreflightEvidence] = None
    preflight_input: Optional[PreflightInput] = None
    preflight_output: Optional[PreflightOutput] = None
    preflight_requires: Optional[PreflightRequires] = None
    record_string_any: Optional[RecordStringAny] = None
    cabin_crew_protocol_record_string_any: Optional[RecordStringAnyClass] = None
    state: Optional[State] = None
    workflow_state: Optional[WorkflowState] = None

File: lib/go/cabincrew/protocol.go
package cabincrew

type CabinCrewProtocol struct {
	AnyMap                           map[string]interface{}  `json:"AnyMap,omitempty"`
	ApprovalRequest                  *ApprovalRequest        `json:"ApprovalRequest,omitempty"`
	ApprovalResponse                 *ApprovalResponse       `json:"ApprovalResponse,omitempty"`
	Artifact                         *Artifact               `json:"Artifact,omitempty"`
	AuditApproval                    *AuditApproval          `json:"AuditApproval,omitempty"`
	AuditArtifact                    *AuditArtifact          `json:"AuditArtifact,omitempty"`
	AuditEngine                      *AuditEngine            `json:"AuditEngine,omitempty"`
	AuditEvent                       *AuditEvent             `json:"AuditEvent,omitempty"`
	AuditGateway                     *AuditGateway           `json:"AuditGateway,omitempty"`
	AuditIntegrity                   *AuditIntegrity         `json:"AuditIntegrity,omitempty"`
	AuditPolicy                      *AuditPolicy            `json:"AuditPolicy,omitempty"`
	AuditWorkflow                    *AuditWorkflow          `json:"AuditWorkflow,omitempty"`
	Decision                         *Decision               `json:"Decision,omitempty"`
	EngineArtifact                   *EngineArtifact         `json:"EngineArtifact,omitempty"`
	EngineInput                      *EngineInput            `json:"EngineInput,omitempty"`
	EngineMeta                       *EngineMeta             `json:"EngineMeta,omitempty"`
	EngineMetric                     *EngineMetric           `json:"EngineMetric,omitempty"`
	EngineOrchestrator               *EngineOrchestrator     `json:"EngineOrchestrator,omitempty"`
	EngineOutput                     *EngineOutput           `json:"EngineOutput,omitempty"`
	GatewayApproval                  *GatewayApproval        `json:"GatewayApproval,omitempty"`
	LLMGatewayPolicyConfig           *LLMGatewayPolicyConfig `json:"LLMGatewayPolicyConfig,omitempty"`
	LLMGatewayRequest                *LLMGatewayRequest      `json:"LLMGatewayRequest,omitempty"`
	LLMGatewayResponse               *LLMGatewayResponse     `json:"LLMGatewayResponse,omitempty"`
	LLMGatewayRule                   *LLMGatewayRule         `json:"LLMGatewayRule,omitempty"`
	MCPGatewayPolicyConfig           *MCPGatewayPolicyConfig `json:"MCPGatewayPolicyConfig,omitempty"`
	MCPGatewayRequest                *MCPGatewayRequest      `json:"MCPGatewayRequest,omitempty"`
	MCPGatewayResponse               *MCPGatewayResponse     `json:"MCPGatewayResponse,omitempty"`
	MCPGatewayRule                   *MCPGatewayRule         `json:"MCPGatewayRule,omitempty"`
	Mode                             *Mode                   `json:"Mode,omitempty"`
	PlanArtifactHash                 *PlanArtifactHash       `json:"PlanArtifactHash,omitempty"`
	PlanToken                        *PlanToken              `json:"PlanToken,omitempty"`
	PreflightEvidence                *PreflightEvidence      `json:"PreflightEvidence,omitempty"`
	PreflightInput                   *PreflightInput         `json:"PreflightInput,omitempty"`
	PreflightOutput                  *PreflightOutput        `json:"PreflightOutput,omitempty"`
	PreflightRequires                *PreflightRequires      `json:"PreflightRequires,omitempty"`
	RecordStringAny                  *RecordStringAny        `json:"Record<string,any>,omitempty"`
	CabinCrewProtocolRecordStringAny *RecordStringAnyClass   `json:"RecordStringAny,omitempty"`
	State                            *State                  `json:"State,omitempty"`
	WorkflowState                    *WorkflowState          `json:"WorkflowState,omitempty"`
}

type ApprovalRequest struct {
	ApprovalID    string              `json:"approval_id"`
	EngineOutput  *RecordStringAny    `json:"engine_output,omitempty"`
	Evidence      []PreflightEvidence `json:"evidence,omitempty"`
	PlanTokenHash *string             `json:"plan_token_hash,omitempty"`
	Reason        string              `json:"reason"`
	RequiredRole  string              `json:"required_role"`
	StepID        string              `json:"step_id"`
	WorkflowID    string              `json:"workflow_id"`
}

// Arbitrary metadata. Optional.
type RecordStringAny struct {
}

type PreflightEvidence struct {
	Hash string `json:"hash"`
	Name string `json:"name"`
	Path string `json:"path"`
}

type ApprovalResponse struct {
	ApprovalID string  `json:"approval_id"`
	Approved   bool    `json:"approved"`
	Approver   *string `json:"approver,omitempty"`
	Reason     *string `json:"reason,omitempty"`
	Timestamp  *string `json:"timestamp,omitempty"`
}

// Canonical artifact interface.
// Defined in schemas/draft/artifact.schema.json
type Artifact struct {
	// Operation to perform with this artifact (create, update, delete, apply, execute, etc).                    
	// Free-form and engine-defined.                                                                             
	Action                                                                                      string           `json:"action"`
	// Type of artifact (file, diff, patch, action, message, etc). Free-form and engine-defined.                 
	ArtifactType                                                                                string           `json:"artifact_type"`
	// Inline content for small artifacts.                                                                       
	// Can be string, object, array, or null.                                                                    
	Body                                                                                        *Body            `json:"body"`
	// Indicates an external data file within the artifact directory.                                            
	BodyFile                                                                                    *string          `json:"body_file,omitempty"`
	// Arbitrary metadata. Optional.                                                                             
	Metadata                                                                                    *RecordStringAny `json:"metadata,omitempty"`
	// MIME type describing content.                                                                             
	MIME                                                                                        string           `json:"mime"`
	// Path, resource, or identifier this artifact applies to. Optional.                                         
	Target                                                                                      *string          `json:"target,omitempty"`
}

type AuditApproval struct {
	ApprovalID   *string `json:"approval_id,omitempty"`
	Approved     *bool   `json:"approved,omitempty"`
	Approver     *string `json:"approver,omitempty"`
	Reason       *string `json:"reason,omitempty"`
	RequiredRole *string `json:"required_role,omitempty"`
}

type AuditArtifact struct {
	Hash *string  `json:"hash,omitempty"`
	Name *string  `json:"name,omitempty"`
	Path *string  `json:"path,omitempty"`
	Role *string  `json:"role,omitempty"`
	Size *float64 `json:"size,omitempty"`
}

type AuditEngine struct {
	EngineID  *string `json:"engine_id,omitempty"`
	Error     *string `json:"error,omitempty"`
	ReceiptID *string `json:"receipt_id,omitempty"`
	Status    *string `json:"status,omitempty"`
}

// Canonical schema for all audit log events.
// Defined in schemas/draft/audit-event.schema.json
type AuditEvent struct {
	Approval                                                                                *AuditApproval  `json:"approval,omitempty"`
	Artifacts                                                                               []AuditArtifact `json:"artifacts,omitempty"`
	// Hash of the previous event in the chain. Allows for ledger-style verification.                       
	ChainHash                                                                               *string         `json:"chain_hash,omitempty"`
	Engine                                                                                  *AuditEngine    `json:"engine,omitempty"`
	// Unique identifier for this audit event.                                                              
	EventID                                                                                 string          `json:"event_id"`
	// Free-form event category.                                                                            
	EventType                                                                               string          `json:"event_type"`
	Gateway                                                                                 *AuditGateway   `json:"gateway,omitempty"`
	IntegrityCheck                                                                          *AuditIntegrity `json:"integrity_check,omitempty"`
	Message                                                                                 *string         `json:"message,omitempty"`
	// Cryptographic binding between a flight-plan and its subsequent take-off.                             
	// Defined in schemas/draft/plan-token.schema.json                                                      
	PlanToken                                                                               *PlanToken      `json:"plan_token,omitempty"`
	Policy                                                                                  *AuditPolicy    `json:"policy,omitempty"`
	Severity                                                                                *Severity       `json:"severity,omitempty"`
	// Cryptographic signature of this event hash.                                                          
	Signature                                                                               *string         `json:"signature,omitempty"`
	// Reference to the key used for signing (e.g. 'engine-key-1', 'orchestrator-key-prod').                
	SignatureKeyRef                                                                         *string         `json:"signature_key_ref,omitempty"`
	// RFC3339 timestamp of when the event occurred.                                                        
	Timestamp                                                                               string          `json:"timestamp"`
	Workflow                                                                                *AuditWorkflow  `json:"workflow,omitempty"`
}

type AuditGateway struct {
	GatewayType    *string `json:"gateway_type,omitempty"`
	Model          *string `json:"model,omitempty"`
	PolicyDecision *string `json:"policy_decision,omitempty"`
	RequestID      *string `json:"request_id,omitempty"`
	Tool           *string `json:"tool,omitempty"`
}

type AuditIntegrity struct {
	ActualPlanToken   *string  `json:"actual_plan_token,omitempty"`
	ArtifactsMatch    *bool    `json:"artifacts_match,omitempty"`
	Differences       []string `json:"differences,omitempty"`
	ExpectedPlanToken *string  `json:"expected_plan_token,omitempty"`
	PlanTokenMatch    *bool    `json:"plan_token_match,omitempty"`
}

// Cryptographic binding between a flight-plan and its subsequent take-off.
// Defined in schemas/draft/plan-token.schema.json
type PlanToken struct {
	// Per-artifact hashes that contributed to this plan token.                                        
	Artifacts                                                                       []PlanArtifactHash `json:"artifacts"`
	// Timestamp when the plan was created (RFC3339).                                                  
	CreatedAt                                                                       string             `json:"created_at"`
	// Engine identity that produced this plan.                                                        
	EngineID                                                                        string             `json:"engine_id"`
	// AI Model identifier used to generate this plan (e.g. 'gpt-4', 'claude-3').                      
	// Required for provenance.                                                                        
	Model                                                                           string             `json:"model"`
	// Engine protocol version used when this plan was produced.                                       
	ProtocolVersion                                                                 string             `json:"protocol_version"`
	// Primary plan token identifier, e.g. SHA256 over all plan artifacts + context.                   
	Token                                                                           string             `json:"token"`
	// Hash of the workspace state when the plan was created.                                          
	WorkspaceHash                                                                   string             `json:"workspace_hash"`
}

type PlanArtifactHash struct {
	Hash string   `json:"hash"`
	Name string   `json:"name"`
	Size *float64 `json:"size,omitempty"`
}

type AuditPolicy struct {
	Decision   *string  `json:"decision,omitempty"`
	Engine     *string  `json:"engine,omitempty"`
	Violations []string `json:"violations,omitempty"`
	Warnings   []string `json:"warnings,omitempty"`
}

type AuditWorkflow struct {
	Mode       *string `json:"mode,omitempty"`
	StepID     *string `json:"step_id,omitempty"`
	WorkflowID *string `json:"workflow_id,omitempty"`
}

type RecordStringAnyClass struct {
}

type EngineArtifact struct {
	Hash string   `json:"hash"`
	Name string   `json:"name"`
	Path string   `json:"path"`
	Role string   `json:"role"`
	Size *float64 `json:"size,omitempty"`
}

// Input delivered via STDIN or CABINCREW_INPUT_FILE.
// Defined in schemas/draft/engine.schema.json
type EngineInput struct {
	AllowedSecrets                                                []string               `json:"allowed_secrets,omitempty"`
	Config                                                        map[string]interface{} `json:"config,omitempty"`
	Context                                                       map[string]interface{} `json:"context,omitempty"`
	ExpectedPlanToken                                             *string                `json:"expected_plan_token,omitempty"`
	// Ephemeral identity token (e.g. OIDC, JWT) for the workload.                       
	// Preferred over static secrets.                                                    
	IdentityToken                                                 *string                `json:"identity_token,omitempty"`
	Meta                                                          EngineMeta             `json:"meta"`
	// Execution mode: 'flight-plan' or 'take-off'.                                      
	Mode                                                          Mode                   `json:"mode"`
	Orchestrator                                                  *EngineOrchestrator    `json:"orchestrator,omitempty"`
	ProtocolVersion                                               string                 `json:"protocol_version"`
	Secrets                                                       map[string]interface{} `json:"secrets,omitempty"`
}

type EngineMeta struct {
	StepID     string `json:"step_id"`
	WorkflowID string `json:"workflow_id"`
}

type EngineOrchestrator struct {
	ArtifactsSalt *string  `json:"artifacts_salt,omitempty"`
	RunIndex      *float64 `json:"run_index,omitempty"`
	WorkspaceHash *string  `json:"workspace_hash,omitempty"`
}

type EngineMetric struct {
	Name  string                 `json:"name"`
	Tags  map[string]interface{} `json:"tags,omitempty"`
	Value float64                `json:"value"`
}

// Output delivered via STDOUT or CABINCREW_OUTPUT_FILE.
// Defined in schemas/draft/engine.schema.json
type EngineOutput struct {
	Artifacts                                         []EngineArtifact `json:"artifacts,omitempty"`
	Diagnostics                                       interface{}      `json:"diagnostics"`
	EngineID                                          string           `json:"engine_id"`
	Error                                             *string          `json:"error,omitempty"`
	Metrics                                           []EngineMetric   `json:"metrics,omitempty"`
	Mode                                              Mode             `json:"mode"`
	// SHA256 hash referencing a plan-token.json file.                 
	PlanToken                                         *string          `json:"plan_token,omitempty"`
	ProtocolVersion                                   string           `json:"protocol_version"`
	ReceiptID                                         string           `json:"receipt_id"`
	// Execution status: 'success' or 'failure'.                       
	Status                                            string           `json:"status"`
	Warnings                                          []string         `json:"warnings,omitempty"`
}

type GatewayApproval struct {
	ApprovalID   *string `json:"approval_id,omitempty"`
	Reason       *string `json:"reason,omitempty"`
	RequiredRole *string `json:"required_role,omitempty"`
}

type LLMGatewayPolicyConfig struct {
	ModelRouting *RecordStringAny `json:"model_routing,omitempty"`
	OnnxModels   []string         `json:"onnx_models,omitempty"`
	OpaPolicies  []string         `json:"opa_policies,omitempty"`
	Rules        []LLMGatewayRule `json:"rules,omitempty"`
}

type LLMGatewayRule struct {
	Action   string           `json:"action"`
	Match    RecordStringAny  `json:"match"`
	Metadata *RecordStringAny `json:"metadata,omitempty"`
}

type LLMGatewayRequest struct {
	Context   *RecordStringAny `json:"context,omitempty"`
	Input     RecordStringAny  `json:"input"`
	Model     string           `json:"model"`
	Provider  *string          `json:"provider,omitempty"`
	RequestID string           `json:"request_id"`
	Source    *string          `json:"source,omitempty"`
	Timestamp string           `json:"timestamp"`
}

type LLMGatewayResponse struct {
	Approval       *GatewayApproval `json:"approval,omitempty"`
	Decision       Decision         `json:"decision"`
	GatewayPayload *RecordStringAny `json:"gateway_payload,omitempty"`
	RequestID      string           `json:"request_id"`
	RewrittenInput *RecordStringAny `json:"rewritten_input,omitempty"`
	RoutedModel    *string          `json:"routed_model,omitempty"`
	Timestamp      string           `json:"timestamp"`
	Violations     []string         `json:"violations,omitempty"`
	Warnings       []string         `json:"warnings,omitempty"`
}

type MCPGatewayPolicyConfig struct {
	OnnxModels  []string         `json:"onnx_models,omitempty"`
	OpaPolicies []string         `json:"opa_policies,omitempty"`
	Rules       []MCPGatewayRule `json:"rules,omitempty"`
}

type MCPGatewayRule struct {
	Action   string           `json:"action"`
	Match    RecordStringAny  `json:"match"`
	Metadata *RecordStringAny `json:"metadata,omitempty"`
}

type MCPGatewayRequest struct {
	Context   *RecordStringAny `json:"context,omitempty"`
	Method    string           `json:"method"`
	Params    *RecordStringAny `json:"params,omitempty"`
	RequestID string           `json:"request_id"`
	ServerID  string           `json:"server_id"`
	Source    *string          `json:"source,omitempty"`
	Timestamp string           `json:"timestamp"`
}

type MCPGatewayResponse struct {
	Approval         *GatewayApproval `json:"approval,omitempty"`
	Decision         Decision         `json:"decision"`
	RequestID        string           `json:"request_id"`
	RewrittenRequest *RecordStringAny `json:"rewritten_request,omitempty"`
	Timestamp        string           `json:"timestamp"`
	Violations       []string         `json:"violations,omitempty"`
	Warnings         []string         `json:"warnings,omitempty"`
}

type PreflightInput struct {
	Context                                                                    *RecordStringAny    `json:"context,omitempty"`
	EngineOutput                                                               RecordStringAny     `json:"engine_output"`
	Evidence                                                                   []PreflightEvidence `json:"evidence,omitempty"`
	Mode                                                                       Mode                `json:"mode"`
	// Cryptographic binding between a flight-plan and its subsequent take-off.                    
	// Defined in schemas/draft/plan-token.schema.json                                             
	PlanToken                                                                  *PlanToken          `json:"plan_token,omitempty"`
	StepID                                                                     string              `json:"step_id"`
	WorkflowID                                                                 string              `json:"workflow_id"`
}

type PreflightOutput struct {
	Decision   Decision           `json:"decision"`
	Requires   *PreflightRequires `json:"requires,omitempty"`
	Violations []string           `json:"violations,omitempty"`
	Warnings   []string           `json:"warnings,omitempty"`
}

type PreflightRequires struct {
	Reason *string `json:"reason,omitempty"`
	Role   *string `json:"role,omitempty"`
}

type WorkflowState struct {
	LastDecision  *string `json:"last_decision,omitempty"`
	PlanTokenHash *string `json:"plan_token_hash,omitempty"`
	State         State   `json:"state"`
	StepID        *string `json:"step_id,omitempty"`
	WorkflowID    *string `json:"workflow_id,omitempty"`
}

type Severity string

const (
	Critical Severity = "critical"
	Debug    Severity = "debug"
	Error    Severity = "error"
	Info     Severity = "info"
	Warning  Severity = "warning"
)

type Decision string

const (
	Allow           Decision = "allow"
	Deny            Decision = "deny"
	RequireApproval Decision = "require_approval"
	Warn            Decision = "warn"
)

// Execution mode: 'flight-plan' or 'take-off'.
type Mode string

const (
	FlightPlan Mode = "flight-plan"
	TakeOff    Mode = "take-off"
)

type State string

const (
	Approved           State = "APPROVED"
	ArtifactsValidated State = "ARTIFACTS_VALIDATED"
	AwaitingApproval   State = "AWAITING_APPROVAL"
	Completed          State = "COMPLETED"
	ExecutionComplete  State = "EXECUTION_COMPLETE"
	Failed             State = "FAILED"
	Init               State = "INIT"
	PlanGenerated      State = "PLAN_GENERATED"
	PlanRunning        State = "PLAN_RUNNING"
	PreFlightRunning   State = "PRE_FLIGHT_RUNNING"
	PreflightComplete  State = "PREFLIGHT_COMPLETE"
	ReadyForTakeoff    State = "READY_FOR_TAKEOFF"
	TakeoffRunning     State = "TAKEOFF_RUNNING"
	TokenCreated       State = "TOKEN_CREATED"
)

// Inline content for small artifacts.
// Can be string, object, array, or null.
type Body struct {
	AnythingArray []interface{}
	AnythingMap   map[string]interface{}
	String        *string
}
